packages/vite/src/node/optimizer/scan.ts

```diff
@@ -606,12 +607,15 @@ function esbuildScanPlugin(
               return externalUnlessEntry({ path: id })
             }
 
-            const namespace = htmlTypesRE.test(resolved) ? 'html' : undefined
-
-            return {
-              path: path.resolve(cleanUrl(resolved)),
-              namespace,
+            if (htmlTypesRE.test(resolved)) {
+              return {
+                path: path.resolve(cleanUrl(resolved)),
+                namespace: 'html',
+              }
             }
+
+            // make esbuild handle js/ts/jsx/tsx
+            // so that tsconfig.json is respected by esbuild
```

Comment:
Would this mean that we let esbuild resolve instead? Wouldn't that possibly diverge from Vite's resolver that we want to use instead?

Maybe we can fix this issue by enabling `experimentalDecorators` by default only for scanning 🤔

Reply:
Hmm true.
It seems esbuild doesn't support reading `tsconfig.json` if the plugin resolved the path (https://github.com/evanw/esbuild/issues/2265).

related code: [the part resolving tsconfig](https://github.com/evanw/esbuild/blob/0a8d2f9cf0ebd32c2ec93e5bd1538cbb2ce148de/internal/resolver/resolver.go#L884-L903), [resolver](https://github.com/evanw/esbuild/blob/0a8d2f9cf0ebd32c2ec93e5bd1538cbb2ce148de/internal/bundler/bundler.go#L861-L977) (`res.Resolve` calls `finalizeResolve` which resolves tsconfig)

Reply:
As much as I'd like to avoid it, maybe reverting and enabling back `experimentalDecorators` for scanning would be the easier way for now 🤔  I think it shouldn't hurt too much since we only need to scan things.

Reply:
Reading https://github.com/microsoft/TypeScript/pull/50820, it seems there's some incompatibilities between the syntax of them unfortunately. 😢
[Example 1 .(private fields)](https://esbuild.github.io/try/#YgAwLjE5LjgALS10YXJnZXQ9ZXNuZXh0AGUAZW50cnkudHMAZXhwb3J0IGRlZmF1bHQgY2xhc3MgRm9vIHsKICBAZm9vCiAgI2ZvbyA9ICdmb28nCn0AAHRzY29uZmlnLmpzb24AewogICJjb21waWxlck9wdGlvbnMiOiB7CiAgICAiZXhwZXJpbWVudGFsRGVjb3JhdG9ycyI6IHRydWUKICB9Cn0), [Example 2 (class expressions)](https://esbuild.github.io/try/#YgAwLjE5LjgALS10YXJnZXQ9ZXNuZXh0AGUAZW50cnkudHMAZXhwb3J0IGNvbnN0IEV4cCA9IEBiYXIgY2xhc3Mge30AAHRzY29uZmlnLmpzb24AewogICJjb21waWxlck9wdGlvbnMiOiB7CiAgICAiZXhwZXJpbWVudGFsRGVjb3JhdG9ycyI6IHRydWUKICB9Cn0) (the error disappears if `experimentalDecorators` is set to `false`)